Source/WebCore/platform/graphics/cg/CGUtilities.h

```diff
@@ -0,0 +1,95 @@
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "FloatRect.h"
+#include "GraphicsTypes.h"
+#include "IntRect.h"
+#include <CoreGraphics/CoreGraphics.h>
+#include <math.h>
+
+namespace WebCore {
+
+inline CGInterpolationQuality cgInterpolationQuality(InterpolationQuality quality)
+{
+    switch (quality) {
+    case InterpolationQuality::Default:
+        return kCGInterpolationDefault;
+    case InterpolationQuality::DoNotInterpolate:
+        return kCGInterpolationNone;
+    case InterpolationQuality::Low:
+        return kCGInterpolationLow;
+    case InterpolationQuality::Medium:
+        return kCGInterpolationMedium;
+    case InterpolationQuality::High:
+        return kCGInterpolationHigh;
+    }
+    ASSERT_NOT_REACHED();
+    return kCGInterpolationDefault;
+}
+
+inline FloatRect cgRoundToDevicePixelsNonIdentity(CGAffineTransform deviceMatrix, FloatRect rect)
+{
+    // It is not enough just to round to pixels in device space. The rotation part of the
+    // affine transform matrix to device space can mess with this conversion if we have a
+    // rotating image like the hands of the world clock widget. We just need the scale, so
+    // we get the affine transform matrix and extract the scale.
+    auto deviceScaleX = hypot(deviceMatrix.a, deviceMatrix.b);
+    auto deviceScaleY = hypot(deviceMatrix.c, deviceMatrix.d);
+
+    CGPoint deviceOrigin = CGPointMake(rect.x() * deviceScaleX, rect.y() * deviceScaleY);
+    CGPoint deviceLowerRight = CGPointMake((rect.x() + rect.width()) * deviceScaleX,
+        (rect.y() + rect.height()) * deviceScaleY);
+
+    deviceOrigin.x = roundf(deviceOrigin.x);
+    deviceOrigin.y = roundf(deviceOrigin.y);
+    deviceLowerRight.x = roundf(deviceLowerRight.x);
+    deviceLowerRight.y = roundf(deviceLowerRight.y);
+
+    // Don't let the height or width round to 0 unless either was originally 0
+    if (deviceOrigin.y == deviceLowerRight.y && rect.height())
+        deviceLowerRight.y += 1;
+    if (deviceOrigin.x == deviceLowerRight.x && rect.width())
+        deviceLowerRight.x += 1;
+
+    FloatPoint roundedOrigin = FloatPoint(deviceOrigin.x / deviceScaleX, deviceOrigin.y / deviceScaleY);
+    FloatPoint roundedLowerRight = FloatPoint(deviceLowerRight.x / deviceScaleX, deviceLowerRight.y / deviceScaleY);
+    return FloatRect(roundedOrigin, roundedLowerRight - roundedOrigin);
+}
+
+inline FloatRect cgRoundToDevicePixels(CGAffineTransform deviceMatrix, FloatRect rect)
```

Comment:
Is there a reason to have these functions inline? Can't we have a source and a header files?

Reply:
Well, it optimizes pretty much to a*b+c type algorithm, so it's kind of in line with other short inline functions..